### PR TITLE
message-here

### PR DESCRIPTION
**Pull request by OmarBeshary**

### CODE REVIEW

File: anabeesh/build.gradle [hunk 1]
```diff
@@ -42,6 +42,10 @@ android {
 }
 
 dependencies {
+
+    compile 'com.google.code.gson:gson:2.8.2'
+    compile 'com.squareup.retrofit2:converter-gson:2.3.0'
```
Comment (Muhammadyoussef) [inline anabeesh/build.gradle]: 1- e7na 3andena file lel dependencies esmo `dependencies.gradle` 7ot feeh el 7agat di w e3mel reference hena

2- `compile` di deprecated, estakhdem `implementation`

File: anabeesh/src/main/java/com/cs18/anabeesh/api/RetrofitApiClient.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+import retrofit2.Retrofit;
+import retrofit2.converter.gson.GsonConverterFactory;
+
+/**
+ Singleton class for Retrofit network connection
+ */
+
+public class RetrofitApiClient {
+    public static final String BASE_URL = "";
```
Comment (Muhammadyoussef) [inline anabeesh/src/main/java/com/cs18/anabeesh/api/RetrofitApiClient.java]: E3mel `//TODO` 3ashan matensash w tseebo keda

File: anabeesh/src/main/java/com/cs18/anabeesh/api/UserRegistrationQueries.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+import retrofit2.Call;
+import retrofit2.http.Body;
+import retrofit2.http.POST;
+
+/**
+ This interface holds all the queries needed to register with Anabeesh's API
+ */
+
+public interface UserRegistrationQueries {
```
Comment (Muhammadyoussef) [inline anabeesh/src/main/java/com/cs18/anabeesh/api/UserRegistrationQueries.java]: Mate3melsh class l kol 7aga lwa7daha, e3mel msln class esmaha `APIService` feeha kol da

File: anabeesh/src/main/java/com/cs18/anabeesh/store/model/login/LogInUserRequest.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+    String getUserEmail() {
+        return userEmail;
+    }
+
+    String getUserPassword() {
+        return userPassword;
+    }
+
+    public void setUserEmail(String userEmail) {
```
Comment (Muhammadyoussef) [inline anabeesh/src/main/java/com/cs18/anabeesh/store/model/login/LogInUserRequest.java]: 7awel telghi mawdou3  set da tamamn, ya emma te3mel builder pattern, ya emma takhod el 7agat di fel constructor awwel marra bas w mate3rafsh teghayaro ba3den

Betefre2 lama tebda2 tet3amel ma3 different threads, lw el object msh immutable msh hayeb2 thread safe

File: anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java [hunk 5]
```diff
@@ -44,9 +55,63 @@ public void onCreate(@Nullable Bundle savedInstanceState) {
 
     @OnClick(R.id.btn_login)
     void login() {
+
+        userEmail = emailEditText.getText().toString();
+        userPassword = passwordEditText.getText().toString();
+
+        if (userEmail.matches("")) {
```
Comment (Muhammadyoussef) [inline anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java]: Ehh el talawos el ana shayfo da 😄, el logic ezay fl activity.

Plus badal mate3mel matches di estakhdem regular expression lel email, e3mel class esmaha textUtil msln feeha el validations di

E2ra f 7aga esmaha java regex

File: anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java [hunk 6]
```diff
@@ -44,9 +55,63 @@ public void onCreate(@Nullable Bundle savedInstanceState) {
...
+        else if (userPassword.matches("")) {
+            // TODO : Handling empty password field .
+        }
+        else if (userPassword.matches("") && userEmail.matches("")) {
+            // TODO : Handling empty e-mail and password field .
+        }
+        else {
+            logInUserRequest = new LogInUserRequest();
+            logInUserRequest.setUserEmail(userEmail);
+            logInUserRequest.setUserPassword(userPassword);
```
Comment (Muhammadyoussef) [inline anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java]: Balash set, ya emma builder ya emma constructor

File: anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java [hunk 7]
```diff
@@ -44,9 +55,63 @@ public void onCreate(@Nullable Bundle savedInstanceState) {
...
+
+            UserLogInQueries userLogInQueries =
+                    RetrofitApiClient.getClient().create(UserLogInQueries.class);
+            Call<LogInUserResponse> loggedInUserResponseCall =
+                    userLogInQueries.getUserLogInResponse(logInUserRequest);
+
+            loggedInUserResponseCall.enqueue(new Callback<LogInUserResponse>() {
```
Comment (Muhammadyoussef) [inline anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java]: El network call kaman fel activity? Fahemni enta shaghal ezay 3ashan ab2 fahem w ana ba3mel review, lw hateshtaghal ay 7aga eshta bas 3arafni

File: anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java [hunk 8]
```diff
@@ -44,9 +55,63 @@ public void onCreate(@Nullable Bundle savedInstanceState) {
...
+                @Override
+                public void onResponse(Call<LogInUserResponse> call, Response<LogInUserResponse> response) {
+                    // valid user name and password
+                    if (response.code() == 200) {
```
Comment (Muhammadyoussef) [inline anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java]: Khalli 200 da constant esmo `SUCCESS` msln 3azhan hanestakhdo keteer, w 3ashan yeb2 readable

File: anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java [hunk 9]
```diff
@@ -44,9 +55,63 @@ public void onCreate(@Nullable Bundle savedInstanceState) {
...
+                        LogInUserResponse userResponseObject = response.body();
```
Comment (Muhammadyoussef) [inline anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java]: Lazem te3mel check lw b `null` walla la, momkem yeb2 null

File: anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java [hunk 10]
```diff
@@ -44,9 +55,63 @@ public void onCreate(@Nullable Bundle savedInstanceState) {
...
+                        String userId = userResponseObject.getUserId();
+                        // move user to home page :
+                        navigateToHomePage();
+                    }
+                    else
+                        // invalid user name or password
+                        if (response.code() == 400) {
```
Comment (Muhammadyoussef) [inline anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java]: Lw el code 400 elnmafrod yedkhol fel `onFailure` asln msh hayeb2 fel `onResponse`

File: anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java [hunk 11]
```diff
@@ -44,9 +55,63 @@ public void onCreate(@Nullable Bundle savedInstanceState) {
...
+                        String errorMessage = response.message();
+                        if (errorMessage.equals("Email Is not found")) {
```
Comment (RxMuhammadyoussef) [inline anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java]: Mayenfa3sh te3mel check kea, plus enta msh me7tag te3mel logic, enta hate3mel show lel message bas

File: anabeesh/src/main/res/values/strings.xml [hunk 12]
```diff
@@ -1,4 +1,5 @@
 <resources>
+
```
Comment (Muhammadyoussef) [inline anabeesh/src/main/res/values/strings.xml]: Avoid adding empty lines 3al fady, w dayman e3mel refactor lel code

**Review by Muhammadyoussef — CHANGES_REQUESTED** [review]

File: anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java (hunk 7 above)
Comment (OmarBeshary) [reply anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java]: sh8al mo2ktn ay 7aga l8ayet lma no3od swa afhm el goz2 dah

File: anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java (hunk 8 above)
Comment (OmarBeshary) [reply anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java]: class lw7do lel constants ? wla globl 3la el class ?

File: anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java (hunk 9 above)
Comment (OmarBeshary) [reply anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java]: done bs law id gy b null * mn el  LogInUserResponse * h3ml eh ?

File: anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java (hunk 10 above)
Comment (OmarBeshary) [reply anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java]: Failure ll network , y3ne el net fsl msln 8er keda la2 msh f el failure 
shof hena keda
https://futurestud.io/tutorials/retrofit-2-catch-server-errors-globally-with-response-interceptor

File: anabeesh/src/main/java/com/cs18/anabeesh/api/RetrofitApiClient.java (hunk 2 above)
Comment (OmarBeshary) [reply anabeesh/src/main/java/com/cs18/anabeesh/api/RetrofitApiClient.java]: done .

File: anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java (hunk 6 above)
Comment (OmarBeshary) [reply anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java]: done . ana 3mltha @Field we keda afdl

File: anabeesh/src/main/res/values/strings.xml (hunk 12 above)
Comment (OmarBeshary) [reply anabeesh/src/main/res/values/strings.xml]: 👍

File: anabeesh/src/main/java/com/cs18/anabeesh/api/UserRegistrationQueries.java (hunk 3 above)
Comment (OmarBeshary) [reply anabeesh/src/main/java/com/cs18/anabeesh/api/UserRegistrationQueries.java]: Class kbeer l retrofit we gowah 2 inner interface ?

File: anabeesh/build.gradle (hunk 1 above)
Comment (OmarBeshary) [reply anabeesh/build.gradle]: wlahy dwrt we ml2tooosh

File: anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java (hunk 5 above)
Comment (OmarBeshary) [reply anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java]: - Done. 3mlt TextValidation Class , we 7teet feh isValidEmail b el regx
- 5leet matches tchof law empty field .

File: anabeesh/src/main/java/com/cs18/anabeesh/api/UserRegistrationQueries.java (hunk 3 above)
Comment (Muhammadyoussef) [reply anabeesh/src/main/java/com/cs18/anabeesh/api/UserRegistrationQueries.java]: Ana asdi interface ya3ni m3lish msh class 7arfeyan

File: anabeesh/build.gradle (hunk 1 above)
Comment (Muhammadyoussef) [reply anabeesh/build.gradle]: Project view > tools > gradle

Aw double shift w e3mel search

File: anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java (hunk 5 above)
Comment (Muhammadyoussef) [reply anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java]: Enta lw 3amalt add l da `(?=\S+$)` fel regex beta3ak msh haye2bal empty space khales, msh me7tag te3mel matches lel " ".

Plus lw 3ayez te3mel check 3ala 7aga empty estakhdem da `TextUtils.isEmpty(string)` di bete3mel check lw empty aw null w betragga3 boolean

File: anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java (hunk 8 above)
Comment (Muhammadyoussef) [reply anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java]: La msh class lwa7do, shoof class monasba leeh (ex. ApiUtil, aw ApiResponse, ay 7aga monasba) w e3melo public static final

File: anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java (hunk 9 above)
Comment (Muhammadyoussef) [reply anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java]: Mayenfa3sh yeb2 el body msh null w el code 200 w el id null, yeb2 akeed enta f api tanya 😂

File: anabeesh/build.gradle (hunk 1 above)
Comment (Muhammadyoussef) [reply anabeesh/build.gradle]: F kol el a7wal lw 3ayez te3mello add keda bera7tK bas knaleeha implementation

File: anabeesh/src/main/java/com/cs18/anabeesh/api/UserRegistrationQueries.java (hunk 3 above)
Comment (OmarBeshary) [reply anabeesh/src/main/java/com/cs18/anabeesh/api/UserRegistrationQueries.java]: done.3mlto el class we gowah el 2 interfaces

File: anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java (hunk 8 above)
Comment (OmarBeshary) [reply anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java]: done . 3mlto public static final

File: anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java (hunk 9 above)
Comment (OmarBeshary) [reply anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java]: tb ma h3ml chick law b null law howa f 200 leeh ?

File: anabeesh/build.gradle (hunk 1 above)
Comment (OmarBeshary) [reply anabeesh/build.gradle]: done . 3mltha zy mnta 3mlha rwsh neek 'D

File: anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java (hunk 5 above)
Comment (OmarBeshary) [reply anabeesh/src/main/java/com/cs18/anabeesh/ui/login/LoginActivity.java]: done 3mltha b 2edy gowa class TextValidation , gowa isValidEmail + isEmpty for any string

**Pull request closed without merging**